woocommerce/woocommerce
# Implement full basic authentication & query string auth without mixing the two (#8256)

Fixes #8256.

This PR does a number of things:
- Versions the REST API into v3 so we can make some larger auth changes. v2 and v1 routes will still work ([mentioned here](https://github.com/woothemes/woocommerce/issues/8256#issuecomment-107682873))
- Fixes basic auth so that it conforms to the Basic auth spec (returns correct headers, etc) instead of the weird hybrid approach present in current versions (and documented more [here and in the rest of the thread](https://github.com/woothemes/woocommerce/issues/8256#issuecomment-107380883))
- Continues to support query string auth, but does so without basic auth getting in the way or the two messing with each other

**Testing Steps**:
- Generate a consumer key and secret if you don't have them already
- Visit /wc-api/v3/products in your browser over https
- See that you are prompted with a basic http auth dialog
- Type in a bogus username/password and see that you get an error echoed to the screen (not a json error)
- Try again and this time provide your correct credentials
- See that the products json is returned
- Visit /wc-api/v3/orders and see that your auth is retained
- Add ?consumer_key=&consumer_secret= to the URL, and fill in your key and secret
- See that you still get a response
- Change consumer_secret to a bogus secret and see a JSON error returned
- Remove the query strings and see that basic auth still works
